torch_np/tests/test_ufuncs_basic.py

```diff
@@ -131,3 +193,140 @@ def test_xy_and_out_broadcast(self, ufunc):
         assert_equal(res_out, res_bcast)
         assert res_out is out
 
+
+dtypes_numeric = [np.int32, np.float32, np.float64, np.complex128]
+
+
+class TestNdarrayDunderVsUfunc:
+    """Test ndarray dunders which delegate to ufuncs, vs ufuncs."""
+
+    @pytest.mark.parametrize("ufunc, op, iop", ufunc_op_iop_numeric)
+    def test_basic(self, ufunc, op, iop):
+        """basic op/rop/iop, no dtypes, no broadcasting"""
+
+        # __add__
+        a = np.array([1, 2, 3])
+        assert_equal(op(a, 1), ufunc(a, 1))
+        assert_equal(op(a, a.tolist()), ufunc(a, a.tolist()))
+        assert_equal(op(a, a), ufunc(a, a))
+
+        # __radd__
+        a = np.array([1, 2, 3])
+        assert_equal(op(1, a), ufunc(a, 1))
+        assert_equal(op(a.tolist(), a), ufunc(a, a.tolist()))
+
+        # __iadd__
+        a0 = np.array([2, 4, 6])
+        a = a0.copy()
+
+        iop(a, 2)     # modifies a in-place
+        assert_equal(a, op(a0, 2))
+
+        a0 = np.array([2, 4, 6])
+        a = a0.copy()
+        iop(a, a)
+        assert_equal(a, op(a0, a0))
+
+    @pytest.mark.parametrize("ufunc, op, iop", ufunc_op_iop_numeric)
+    @pytest.mark.parametrize("other_dtype", dtypes_numeric)
+    def test_other_scalar(self, ufunc, op, iop, other_dtype):
+        """Test op/iop/rop when the other argument is a scalar of a different dtype."""
+        a = np.array([1, 2, 3])
+        b = other_dtype(3)
+
+        if ufunc in no_complex and issubclass(other_dtype, np.complexfloating):
+            pytest.skip(f'{ufunc} does not accept complex.')
+
+        # __op__
+        result = op(a, b)
+        assert_equal(result, ufunc(a, b))
+        assert result.dtype == np.result_type(a, b)
```

Comment:
OK, this (and similar assertion on `result.dtype`) is not always right. Will fix separately.

Reply:
xfail for now in the last commit. This needs fully fledged type promotion, including "weak promotion" from scalars. Will circle back to it after np.random and numpy test suite are ported.